app/provider_details/provider_service.py

```diff
@@ -36,42 +42,77 @@ def init_app(
+                # This exception message is more detailed than the messages below.
                 raise InvalidProviderException(
-                    f'provider strategy {provider_selection_strategy.get_label()} could not find a suitable provider'
+                    f"Could not determine a provider using strategy {provider_selection_strategy.get_label()}."
                 )
+        else:
+            # Do not use any other criteria to determine the provider.  See issue 944.
+            provider = None
+
+        if provider is None:
+            if provider_id:
+                raise InvalidProviderException(f"The provider {provider_id} could not be found.")
+            raise InvalidProviderException("Could not determine a provider.")
+        elif not provider.active:
+            raise InvalidProviderException(f"The provider {provider.display_name} is not active.")
 
+        logger.debug("Returning provider: %s", "None" if (provider is None) else provider.display_name)
         return provider
 
     @staticmethod
     def _get_template_or_service_provider_id(notification: Notification) -> Optional[str]:
+        """
+        Return a primary key (UUID) for an instance of ProviderDetails using this criteria:
+            1. Use the notification template's provider_id first.
+            2. Use the notification service's provider_id if the template's provider_id is null.
+
+        Return None if neither criterion yields a provider ID.
+        """
+
+        # The template provider_id is nullable foreign key (UUID).
+        # TODO - The field is nullable, but what does SQLAlchemy return?  An empty string?
+        # Testing for None broke a user flows test.
         if notification.template.provider_id:
+            logger.debug("Found template provider ID %s", notification.template.provider_id)
             return notification.template.provider_id
 
-        service_provider_id = {
-            NotificationType.EMAIL: notification.service.email_provider_id,
-            NotificationType.SMS: notification.service.sms_provider_id
-        }[NotificationType(notification.notification_type)]
+        # A template provider_id is not available.  Try using a service provider_id, which might also be None.
+        if notification.notification_type == NotificationType.EMAIL.value:
+            logger.debug("Service provider e-mail ID %s", notification.service.email_provider_id)
+            return notification.service.email_provider_id
+        elif notification.notification_type == NotificationType.SMS.value:
+            logger.debug("Service provider SMS ID %s", notification.service.sms_provider_id)
+            return notification.service.sms_provider_id
 
-        if service_provider_id:
-            return service_provider_id
+        # TODO - What about letters?  That is the 3rd enumerated value in NotificationType
+        # and Notification.notification_type.
```

Comment:
Ticket?

Reply:
#957

Reply:
Can you please add the ticket number to the TODO?

Reply:
Added